### PR TITLE
DI error responses

### LINKED ISSUE #1928: Display error details in synchronous DI responses
### Context

Follow-up of #1883.

Once #1920 gets merged, we should further adjust the error handling to include it in the DI responses, preferably in the `value` field: hex string of encoded error type. See also the comment in `trusted_call.rs`.

It should be at least tested via DI-sample. 

---
:heavy_check_mark: Please set appropriate **labels** and **assignees** if applicable.

### PR DESCRIPTION
**Pull request by kziemianek**
* Feeds `WorkerRpcReturnValue` `value` with tee-worker error in case of STF errors.
* Adds standarized Response struct

fixes: #1928

### CODE REVIEW

File: tee-worker/app-libs/stf/src/trusted_call.rs
Comment (kziemianek) [inline tee-worker/app-libs/stf/src/trusted_call.rs]: Can be renamed to `call_id`. I think it's worth to abstract it a little bit and make `stf` less req/rpc aware.

File: tee-worker/core-primitives/stf-executor/src/executor.rs
Comment (kziemianek) [inline tee-worker/core-primitives/stf-executor/src/executor.rs:142]: @Kailai-Wang I think we can do the same to successful TOPs. TOPs can return result value that implements scalec's `Encode` and in case `req_has` is present encode it here and put it in `ExecutedOperation`. What do you think ?

This could lead us to removal of `https://github.com/litentry/litentry-parachain/blob/055b19d358089b0ed2f9571a23badbd82c638f6d/tee-worker/app-libs/stf/src/trusted_call.rs#L280`

File: tee-worker/core-primitives/stf-executor/src/error.rs
Comment (arieljbirnbaum) [inline tee-worker/core-primitives/stf-executor/src/error.rs]: Doesn't it need to derive `TypeInfo`?

File: tee-worker/core-primitives/stf-executor/src/lib.rs
Comment (arieljbirnbaum) [inline tee-worker/core-primitives/stf-executor/src/lib.rs:70]: Two questions:

1. This is now isomorphic to `(H256, Vec<OpaqueCall>, RpcResponseValue, boolean)`; could/should we factor that out? (can be a named enum instead of a naked boolean ofc)
2. Is the success/failure value not included already (albeit encoded) in the `RpcResponseValue`?

File: tee-worker/ts-tests/integration-tests/identity-direct-invocation.test.ts
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/integration-tests/identity-direct-invocation.test.ts]: :nail_care: for easier debugging, either go with

```
assert.equal(v.type, 'LinkIdentityFailed')
```

or print out the actual type in the message:

```
assert.isTrue(v.isLinkIdentityFailed, `expected LinkIdentityFailed, received ${v.type} instead`)
```

File: tee-worker/core-primitives/stf-executor/src/error.rs
Comment (kziemianek) [reply tee-worker/core-primitives/stf-executor/src/error.rs]: Nothing is going to consume generated metadata so we don't need it.

File: tee-worker/core-primitives/stf-executor/src/lib.rs
Comment (kziemianek) [reply tee-worker/core-primitives/stf-executor/src/lib.rs:70]: 1. I prefer to keep it as it is because i like this top level distinction, it feels natural that you have success and failure execution status. Additionaly there is high possibility that in the future it would become nonisomorphic again.
2. even if it is, its hidden behind encoded bytes so there is no easy and cheap way to get access to it.

File: tee-worker/app-libs/stf/src/lib.rs
Comment (Kailai-Wang) [inline tee-worker/app-libs/stf/src/lib.rs:214]: Not now - but maybe later

File: tee-worker/core-primitives/top-pool-author/src/author.rs
Comment (Kailai-Wang) [inline tee-worker/core-primitives/top-pool-author/src/author.rs]: I've been struggling if I should use singular or plural, but in the end I chose singular because a `Vec` is a singular 😂  it can have many elements but I refer `Vec` as the `value` (and it should be `response` not `responses` - that's my typo).

What do you think?

File: tee-worker/app-libs/stf/src/trusted_call.rs
Comment (Kailai-Wang) [reply tee-worker/app-libs/stf/src/trusted_call.rs]: Yea I like the name `call_id` or simply `id` or `identifier`.
In my knowledge, `id` could mean `index` too which might lead to confusion as it is _normally_ implemented as an integer.

File: tee-worker/core-primitives/stf-executor/src/lib.rs
Comment (Kailai-Wang) [inline tee-worker/core-primitives/stf-executor/src/lib.rs]: Seems we'll never get `None`? How about a plain `H256`?

File: tee-worker/core-primitives/stf-executor/src/error.rs
Comment (Kailai-Wang) [inline tee-worker/core-primitives/stf-executor/src/error.rs]: do you think it's better to go to `trusted_call_rpc_response.rs`?

Comment (Kailai-Wang) [reply tee-worker/core-primitives/stf-executor/src/error.rs]: And shall we always position `req_ext_hash` as the last field to be consistent with other `*Response` structs?

File: tee-worker/core-primitives/stf-executor/src/executor.rs
Comment (Kailai-Wang) [reply tee-worker/core-primitives/stf-executor/src/executor.rs:142]: Hmm I'm not sure if I get you. You can do this for `ErrorResponse` because the failed TOP has the error of the same type. The success TOP would have different Response types even though they are all encodable. How would you write the `Ok(res)` here? if you mean `res` is already encoded, I don't see much difference from the current impl..

**Review by Kailai-Wang — COMMENTED** [review]
Looks good in general

File: tee-worker/core-primitives/stf-executor/src/executor.rs
Comment (kziemianek) [reply tee-worker/core-primitives/stf-executor/src/executor.rs:142]: They can be wrapped into enum (just like `StfError`) and handled in the same way.

File: tee-worker/core-primitives/top-pool-author/src/author.rs
Comment (kziemianek) [reply tee-worker/core-primitives/top-pool-author/src/author.rs]: If this is intentional then it's fine, i'll revert 🙃

File: tee-worker/core-primitives/stf-executor/src/error.rs
Comment (kziemianek) [reply tee-worker/core-primitives/stf-executor/src/error.rs]: If we make `Response` generic then it will be enforced
```
#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
pub(crate) struct Response<T: Encode> {
	pub req_ext_hash: H256,
	pub value: T,
}
```

File: tee-worker/core-primitives/stf-executor/src/executor.rs
Comment (kziemianek) [reply tee-worker/core-primitives/stf-executor/src/executor.rs:142]: @Kailai-Wang please check this [commit](https://github.com/litentry/litentry-parachain/commit/84de0ee5af337a9039c299828e719dd3e0eef66e)

File: tee-worker/app-libs/stf/src/trusted_call.rs
Comment (kziemianek) [reply tee-worker/app-libs/stf/src/trusted_call.rs]: I'll use `identifier`

File: tee-worker/core-primitives/top-pool-author/src/author.rs
Comment (Kailai-Wang) [reply tee-worker/core-primitives/top-pool-author/src/author.rs]: Thanks! Can you do me a favour and remove the `s` in the middle? `rpc_responses_value ` => `rpc_response_value `

File: tee-worker/app-libs/stf/src/trusted_call.rs
Comment (arieljbirnbaum) [inline tee-worker/app-libs/stf/src/trusted_call.rs:790]: I think we need to keep this here for compatibility until clients (i.e. IDHub) are no longer relying on it to detect the response, no?

File: tee-worker/bin/integritee-cli
Comment (arieljbirnbaum) [inline tee-worker/bin/integritee-cli:1]: Is this supposed to be committed? o_O

File: tee-worker/ts-tests/integration-tests/substrate_ii_identity.test.ts
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/integration-tests/substrate_ii_identity.test.ts]: Is this check passing? If so, please disregard my previous comment about backwards compatibility, I must have misunderstood the code :)

File: tee-worker/bin/integritee-cli
Comment (kziemianek) [reply tee-worker/bin/integritee-cli:1]: No, these binaries were renamed in the meanwhile so that's why they appeared in commit. Thanks!

File: tee-worker/app-libs/stf/src/trusted_call.rs
Comment (kziemianek) [reply tee-worker/app-libs/stf/src/trusted_call.rs:790]: Marked pull request as breaking.
The clients will need to adapt, sadly there is no backward compatibility guaranteed :(

File: tee-worker/ts-tests/integration-tests/substrate_ii_identity.test.ts
Comment (kziemianek) [reply tee-worker/ts-tests/integration-tests/substrate_ii_identity.test.ts]: Tests pass 🙃

File: tee-worker/bin/integritee-cli
Comment (Kailai-Wang) [inline tee-worker/bin/integritee-cli:1]: We don't need to commit this

File: tee-worker/bin/integritee-service
Comment (Kailai-Wang) [inline tee-worker/bin/integritee-service:1]: Same as above

File: tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts
Comment (Kailai-Wang) [inline tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts]: I think the name `Response` is a bit too generic, do you have a better one?

**Review by Kailai-Wang — COMMENTED** [review]
Since your `Response` always contains `req_ext_hash` as the first field, do you need to change the logic in `RpcResponder::update_status_event`?

Actually if you don't change that I don't expect the two-step trustedCalls to work 🤔

File: tee-worker/app-libs/stf/src/trusted_call.rs
Comment (arieljbirnbaum) [reply tee-worker/app-libs/stf/src/trusted_call.rs:790]: Well if the tests are passing that seems to indicate that the Old Way (ie listening for the response on the parachain events) still works, no? :3

File: tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts
Comment (kziemianek) [reply tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts]: @Kailai-Wang what do you think about `TrustedOperationResult` ?

**Comment by kziemianek** [comment]
> Since your `Response` always contains `req_ext_hash` as the first field, do you need to change the logic in `RpcResponder::update_status_event`?
> 
> Actually if you don't change that I don't expect the two-step trustedCalls to work 🤔

Should work, at least tests pass. 

There is `TrustedCallResult::Streamed` which is represended by encoded `true` value so it should work as before.

```
			// true means that there are more results to come, see rpc_responder
			Self::Streamed => true.encode(),
```

How is this related to `req_ext_hash` ?

**Comment by Kailai-Wang** [comment]
The problem is the `rpc_response_value`, which contains the req_ext_hash in the new impl. And this code snippet:

```
		// Litentry:
		// a trick to use `result.value` as the flag to forcily watch this connection.
		// If it's `true.encode()` it means we have streamed trustedCalls and there's more to come
		// TODO: how about relying on req_ext_hash? See comment in `trusted_call.rs`
		let do_watch = continue_watching(&status_update)
			|| (result.value == true.encode()
				&& matches!(
					result.status,
					DirectRequestStatus::TrustedOperationStatus(
						TrustedOperationStatus::InSidechainBlock(_),
						_
					)
				));
```

It seems `result.value == true.encode()` can never be true here as `value` is actually encoded `Response`

File: tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts
Comment (Kailai-Wang) [reply tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts]: Maybe `TrustedOperationResponse`?  Another dumb question is why we use `Vec<u8>` for `value`, not `TrustedCallResult`? It seems in code that's the only source for encoding

Comment (kziemianek) [reply tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts:273]: I think because for some cases the `value` can hold encoded string

```
	pub fn from_error_message(error_msg: &str) -> Self {
		RpcReturnValue {
			value: error_msg.encode(),
			do_watch: false,
			status: DirectRequestStatus::Error,
		}
	}
```

Comment (kziemianek) [reply tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts]: In some cases it can hold encoded string error 

```
	pub fn from_error_message(error_msg: &str) -> Self {
		RpcReturnValue {
			value: error_msg.encode(),
			do_watch: false,
			status: DirectRequestStatus::Error,
		}
	}
```

Comment (Kailai-Wang) [reply tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts]: I see, thanks. So I need to decode it to different types based on the `status` field in the `RpcReturnValue`?

**Comment by kziemianek** [comment]
> Since your `Response` always contains `req_ext_hash` as the first field, do you need to change the logic in `RpcResponder::update_status_event`?
> 
> Actually if you don't change that I don't expect the two-step trustedCalls to work 🤔

That's true. 
I though we have test case covering this.

File: tee-worker/app-libs/stf/src/trusted_call.rs
Comment (kziemianek) [reply tee-worker/app-libs/stf/src/trusted_call.rs:790]: In terms of response channel - yes it works as before. 
Currently client can:

- in case of II listen to parachain events
- in case of DI wait for response on WS or listen to parachain events

Comment (kziemianek) [reply tee-worker/app-libs/stf/src/trusted_call.rs:790]: By saying there is no backward compatibility i meant that RPC has slightly changed.

File: tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts
Comment (kziemianek) [reply tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts]: Yes, not very convenient.

**Comment by Kailai-Wang** [comment]
> > Since your `Response` always contains `req_ext_hash` as the first field, do you need to change the logic in `RpcResponder::update_status_event`?
> > Actually if you don't change that I don't expect the two-step trustedCalls to work 🤔
> 
> That's true. I though we have test case covering this.

The original web3 verification sample covered this.
But that's not the case anymore after #1989 is merged.

The two-step trustedcalls are:
- web2 verification
- vc request
So we need to have one of them to cover the case :(

**Comment by kziemianek** [comment]
> > > Since your `Response` always contains `req_ext_hash` as the first field, do you need to change the logic in `RpcResponder::update_status_event`?
> > > Actually if you don't change that I don't expect the two-step trustedCalls to work 🤔
> > 
> > 
> > That's true. I though we have test case covering this.
> 
> The original web3 verification sample covered this. But that's not the case anymore after #1989 is merged.
> 
> The two-step trustedcalls are:
> 
> * web2 verification
> * vc request
>   So we need to have one of them to cover the case :(

True, i'm working on adding it.

**Comment by kziemianek** [comment]
@Kailai-Wang there is already test verifying linking of web2 identity:
https://github.com/litentry/litentry-parachain/blob/385c54864e3e45f18621466e75c22806f3dad0f1/tee-worker/ts-tests/integration-tests/di_substrate_identity.test.ts#L48

File: tee-worker/core-primitives/stf-executor/src/executor.rs
Comment (Kailai-Wang) [inline tee-worker/core-primitives/stf-executor/src/executor.rs]: Hmm would you like to rename this to `TrustedOperationResponse` or `TrustedOperationResult`?

File: tee-worker/core-primitives/stf-executor/src/executor.rs
Comment (Kailai-Wang) [inline tee-worker/core-primitives/stf-executor/src/executor.rs:156]: Shall we add a comment regarding this special handling?

**Review by Kailai-Wang — APPROVED** [review]
Thanks!

**Pull request merged**
